### ISSUE TITLE
JaxbAnnotationModule Incorrectly Serializes Object as Ref Only In some Cases

### ISSUE THREAD

**Issue by GonenI** [FasterXML/jackson-modules-base#46]
When using Jackson with JaxbAnnotationModule , if objects with similar ids appears in two different fields of an outer object , it is sometimes incorrectly serialized as a ref only, instead of as a full object
Case 1:
 The Inner object class has both @XMLID and @XMLElement properties
 One of the inner objects appear in a list field of the outer object, and another in another field
 Unit Test to show problem:

class HasID
{
	String id;
	String name;
    @XmlID 
    @XmlElement
    public String getId() {
        return id;
    }
    @XmlElement
    public String getName() { return name; }
}

@XmlRootElement
class HasIDList
{
	List<HasID> elements = null;
    @XmlElement
    public List<HasID> getElements( ) { return elements; }
    public void setElements( List<HasID> elements ) { this.elements = elements; }
    HasID parent;
    @XmlElement 
    public HasID getParent() { return parent; }
}

public class SerializeAsFullObjectTest{

	@Test
	public void test() throws Exception {
		ObjectMapper mapper = new ObjectMapper();
		JaxbAnnotationModule jaxbAnnotationModule = new JaxbAnnotationModule();
		mapper.registerModule(jaxbAnnotationModule);

		HasID hasID = new HasID();   
		hasID.id = "1"; hasID.name="name1";
		HasID hasID2 = new HasID();
		hasID2.id = "1"; hasID2.name="name1";
		
		HasIDList idList = new HasIDList();
		idList.setElements(Arrays.asList(hasID,hasID2));
		idList.parent = hasID2;
		
		Assert.assertEquals("{\"elements\":[{\"id\":\"1\",\"name\":\"name1\"},{\"id\":\"1\",\"name\":\"name1\"}],\"parent\":{\"id\":\"1\",\"name\":\"name1\"}}", mapper.writeValueAsString(idList));
	}
}
 
The problem only appears when
 1) Using JaxbAnnotations, with @XmlID and ( @XmlElement or @XmlAttribute)

This is a real ( non academic ) problem in a current production project.

**Comment by GonenI** [comment]
And thanks for all your work on this great project. I have checked this with 2.0, 2.5 and 2.94. I have not verified with 2.9.5

**Comment by cowtowncoder** [comment]
@GonenI Thanks. 2.9.4 is fine, just wanted to rough verification. Sometimes we get reproduction for old enough versions where there is reasonable chance of a fox (like 2.5 or earlier).

**Comment by cowtowncoder** [comment]
Ok. With a bit more investigation, this is unfortunately one difference between Jackson handling of Object Identity, and that of JAXB. The way JAXB expects `XMLId` to be used is different from the way Jackson annotations work, and mapping in this case does not work the way you would expect.

It would actually be possible to make this work the way you want, but with Jackson annotations, I think. In Jackson, `@JsonIdentityInfo` can be applied either to classes (so that all values of that type will write value once, then use references), or to properties (in which case only references through that property do that).
This is different from JAXB where id property is annotated and not class (containing it) or property that refers to instances. What Jackson does, in attempting to map it to an approximate is to take an existence of `@XmlId` to infer existence of `@JsonIdentityInfo` for that type.

So... if it was possible to use Jackson annotations, you would remove `@XmlId`, and add:

```java
        @JsonIdentityInfo(generator = ObjectIdGenerators.PropertyGenerator.class, property="id")
        public List<HasID> getElements() { return elements; }
```

to `elements` property. This binds id to `id` property, but only enables identity handling for scope of references via this property, not including value referenced through `parent`.

I don't think Jackson's handling of `@XmlId` can be changed to work the way expected here, unfortunately.

**Comment by GonenI** [comment]
Thanks. So just checking if I understand - with the JaxB Annotations module, XMLID behaves as a global directive for the class , and serializing as ref values cannot be restricted only to fields containing @XMLIDREF ?

**Comment by cowtowncoder** [comment]
@GonenI Correct. So Jackson can model the concept, but defaults are different: in JAXB, just existence of XmlID does not automatically enable reference handling. But due to the way Jackson maps JAXB annotations to behavior its native annotations have (in which class either has `@JsonTypeInfo` or not; if not, property may further have it, or override), behavior of `@XmlID` does not quite match JAXB behavior.

I am not sure if it would be possible to improve this; I can see how this is problematic. I am open to ideas -- the main challenge is that internal handling logic is not easy to change.
